### PR TITLE
Use the document base url when resolving form action URLs

### LINKED ISSUE #10580: Use the document base url when resolving form action URLs
According to [the specification](https://html.spec.whatwg.org/multipage/forms.html#attributes-for-form-submission), we need to be [using the base URL](https://html.spec.whatwg.org/multipage/infrastructure.html#parse-a-url). This means calling `document.base_url()` instead of `window.get_url()` as the code currently does (specifically, `make_url_or_base_getter`, which is used for the `action` and `formaction` attributes in elements like `HTMLFormElement`).

Code: `components/script/dom/macros.rs`
Test:
```html
<base href="resources/"></base>
<form action="target.html">
```
Visit this from a web server (eg. `python -m SimpleHTTPServer`) with a `resources/` directory that contains a `target.html` file that isn't empty.

There does not appear to be an automated test for this yet, so we'll need to [write a  test](https://github.com/servo/servo/blob/master/tests/wpt/README.md#writing-new-tests)! It belongs in `tests/wpt/metadata/html/semantics/forms/the-form-element/`. Specifically, it should have an iframe that contains a form with a base url, and the form should be submitted via JavaScript (`form.submit()`). The resulting page should notify the parent page that it was successful (`window.parent.some_success_function()`).

### PR DESCRIPTION
**Pull request by mylainos**
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy --faster` does not report any errors
- [X] These changes fix #10580

The test fail with `TIMEOUT`.
On screen `Unknown content type (application/json).` is displayed, the target of the form isn't recognize as a html file. How can I fix this?

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/11219)

### CODE REVIEW

File: resources/base-url-target.html [hunk 1]
```diff
@@ -0,0 +1,6 @@
+<!DOCTYPE html>
```
Comment (jdm) [inline resources/base-url-target.html]: This should be in `tests/wpt/web-platform-tests/html/semantics/forms/the-form-element/resources/`, not servo's toplevel resources directory :)

**Comment by Mylainos** [comment]
The function `t.step_func_done();` is executed but the test fail (TIMEOUT).

**Comment by Ms2ger** [comment]
`step_func_done` creates a function that finishes the test when called; you could use `var success = t.step_func_done();`

**Comment by Mylainos** [comment]
Everything good now :smile:

File: components/script/dom/macros.rs [hunk 2]
```diff
@@ -2,6 +2,8 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
+use dom::document::Document;
```
Comment (KiChjang) [inline components/script/dom/macros.rs]: Is this actually used?

Comment (Mylainos) [inline components/script/dom/macros.rs]: No, I'll remove it.

**Comment by Mylainos** [comment]
Need changes ?

**Comment by Ms2ger** [comment]
Reviewed 1 of 3 files at r2, 1 of 4 files at r3, 1 of 3 files at r4, 3 of 3 files at r7, 1 of 1 files at r8, 1 of 1 files at r9.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[components/script/dom/macros.rs, line 102 \[r9\]](https://reviewable.io:443/reviews/servo/servo/11219#-KJGR7kQd8na-juJtA0M:-KJGR7kQd8na-juJtA0N:1668775734) ([raw file](https://github.com/servo/servo/blob/91ecb2f4e04ce7595a23c0b6b0cc4f1534959e41/components/script/dom/macros.rs#L102)):*
> ```Rust
>             if url.is_empty() {
>                 let document = document_from_node(self);
>                 DOMString::from(document.r().base_url().into_string())
> ```

Drop the `.r()`

---

*[components/script/dom/macros.rs, line 102 \[r9\]](https://reviewable.io:443/reviews/servo/servo/11219#-KJGUvqGqXBpKIXUmF0y:-KJGUvqGqXBpKIXUmF0z:-887624630) ([raw file](https://github.com/servo/servo/blob/91ecb2f4e04ce7595a23c0b6b0cc4f1534959e41/components/script/dom/macros.rs#L102)):*
Reading the spec, I find:

> The action IDL attribute must reflect the content attribute of the same name, except that on getting, when the content attribute is missing or its value is the empty string, the element's node document's *URL* must be returned instead.

Note: "URL", not "base URL".

I'm not sure what @jdm had in mind, but this doesn't seem to be it.

r? @jdm

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11219)*

**Comment by jdm** [comment]
Ok, the specification is oddly byzantine here. The getter should actually reflect the content value (ie. not return a resolved URL at all), except when it's empty, in which case it should return the document URL (not base URL). HOWEVER, when we use the action value during the form submission process ([steps 8-10](https://html.spec.whatwg.org/multipage/forms.html#concept-form-submit)), we need to resolve the value obtained from the getter against the document's base URL.

**Comment by jdm** [comment]
Sorry, for the incorrect original description in #10580. What we should do now:
* replace each `make_url_or_base_getter` for `action` and `formaction` with a new `make_string_or_document_url` macro that return's the document's URL as a string if the attribute value is empty
* modify `HTMLFormElement::submit` to use the actual document base URL in `base`, and make step 8 use the document's non-base URL instead

**Comment by Mylainos** [comment]
Ok, I will modify this.

**Comment by jdm** [comment]
@Mylainos Are you still planning to finish this?

**Comment by Mylainos** [comment]
Yes !

**Comment by jdm** [comment]
Closing due to inaction. If you continue to work on this, please feel free to reopen this PR.

**Pull request closed without merging**